Add a json reporter

Tummerhore opened the pull request:
Fixes #1814. The implementation is relatively simple and copies most of the code from the html reporter. Probably this can be refactored at some point so that there is less duplicated logic. But it works for now.

nicojs commented on packages/core/src/reporters/json-reporter.ts:
```diff
@@ -0,0 +1,62 @@
+export const RESOURCES_DIR_NAME = 'strykerResources';
+
+export default class JsonReporter implements Reporter {
+  private _baseDir!: string;
```
Please use `_baseDir: string | undefined` here.

nicojs commented on packages/core/src/reporters/json-reporter.ts:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  private async cleanBaseFolder(): Promise<void> {
+    await HtmlReporterUtil.deleteDir(this.baseDir);
```
I don't think we should clean the base dir here. This is by default the directory that the HTML reporter is also writing to. 

 I think you can simply remove this here.

nicojs reviewed (changes requested):
Thanks a lot for this PR! Could you maybe also add some unit tests?

nicojs commented on packages/core/src/reporters/json-reporter.ts:
```diff
@@ -0,0 +1,62 @@
+    await this.cleanBaseFolder();
+    let indent = DEFAULT_INDENTION;
+    if (this.options.jsonReporter && this.options.jsonReporter.spacing) {
+      indent = this.options.jsonReporter.spacing;
```
In hind sight, I think this feature is a bit expensive for what it does. Expensive in the sense that we need to document and maintain it. In practise, the Json reporter would not be used to show the report to a human (i.e. open in an editor).

Is there a specific reason you would want this feature?

Tummerhore replied:
Do you only mean the indention option or the whole json reporter in general? I think that a json reporter should definitely be supported since there is currently no reporter for easy data exchange. If you mean the indention option only, you are probably right. I just added it because it was a very easy addition and is probably wanted. :)

Tummerhore replied:
I've now removed the indention option and use the default value (0 = no indention) as a static constant.

Tummerhore commented:
Ok, finally found some time to fix the remarks: Added some unit tests, removed the indention option and adjusted some things you commented on in the analog pr #2601 which were also present in my pr. I've also added the option for a custom filename besides the custom path.

nicojs reviewed (approved):
This looks great overall! I've got one more remark, please take a look if you have some time. If not, I can merge it in and change it before doing a release. Your choice 👍

nicojs commented:
Thanks! 👍

dance-cmdr commented:
Many thanks for the feature @Tummerhore!
